README.md
# PyVis
PyVis is a data visualization tool that is an extension of [pybaseball](https://github.com/jldbc/pybaseball).

## Getting Started
See below on how to start using PyVis.

### Prerequisites
First, you need to install `pybaseball` by typing the following into the command line: `pip install pybaseball`. If you don't have `pip`, install it [here](https://pip.pypa.io/en/stable/installing/).

It is recommended that you replace any local files installed by `pip` with the files from the `pybaseball` repo, as some functions in `PyVis` use updated functions that have not been updated in the `pip` installation.

## Installation
Currently, you cannot install PyVis via `pip`. In the future, that will be the case. For now, download all the Python files in the PyVis folder into a local folder on your computer.

## How PyVis Works
See [Explaining PyVis.md](https://github.com/kunalcsc630/pyvis/blob/master/pyvis/Explaining%20PyVis.md) to learn more about the capabilities of PyVis. Have fun visualizing!

## Contributing
See [CONTRIBUTING.md](https://github.com/kunalcsc630/pyvis/blob/master/CONTRIBUTING.md) to learn how you can contribute to PyVis.

## License
The PyVis project is under the MIT License. See [LICENSE.md](https://github.com/kunalcsc630/pyvis/blob/master/LICENSE) for more.

## Acknowledgments
* [Dr. Nicholas Zufelt](https://github.com/nzufelt)
* [Mr. James LeDoux](https://github.com/jldbc)
* [Ms. Billie Thompson](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2)
* [CSC630 Classmates](nzufelt.github.io/open_source_movement_csc630/)
* Additional references available [here](https://github.com/kunalcsc630/pyvis/blob/master/REFERENCES.md)


CHANGELOG.md
# Change Log

## Jan 26
I've updated my README to be more like a README for an actual open source repo. I also looked more into `pybaseball` and discovered that the function works if you call it with specific syntax. I'm also noticing that I'm going to need to spend more time understanding the inner workings of `pybaseball` before I can begin working on my own open source library.

## Jan 28
Over the weekend, I worked on understanding `pybaseball`. I now have a fairly good idea of how it works. Now, I'll need to start working on my repo and writing improvements to it.

## Jan 29
In class today, I realized I'll need to learn the inner workings of `matplotlib` because there are so many different things you can do with it. That way, I'll know what kinds of graphs I want to quickly generate in my library. That is what I'll work on over the next two days.

## Jan 30
I'm getting a bit confused why the `statcast` function runs for certain dates and for the exact same dates, doesn't run. I am going to look into how the `statcast` function picks up the customized CSV file from [Baseball Savant](https://baseballsavant.mlb.com/statcast_search). Then, I'll get back to understanding `matplotlib`.

## Feb 2
I tried to figure out why I can't obtain the link to the CSV file. Eventually, Baseball Savant was put under maintenance, so I couldn't work any further on that. So, I started working on PyVis.

## Feb 6
I continued trying to figure out the link to the CSV file. I asked Dr. Z and he said to use [Selenium](http://selenium-python.readthedocs.io/) to simulate the `onclick` event to download the CSV. If it doesn't take too long, I'm going to use Selenium. Otherwise, I'm going to deal with the Statcast search issues because I need to make more progress on my own library.

## Feb 7
Today, I created my first function for PyVis called `hit_breakdown`, which gives you information regarding the events happening at a batter's plate appearance. It can take a singular day or a range of dates (in a certain format for the latter) and visualize the breakdown for the user via a pie chart. For aesthetic purposes, I decided to eliminate the tiny slices in the pie chart where the event happened less than 2% of the time and merged all of those events into a category called 'other'.

## Feb 8
I added a new function today called `pitch_breakdown`, which does what `hit_breakdown` does but for the type of pitch thrown.

## Feb 9
I compiled a list of other ideas I want to implement and posted that list as an issue. Over the weekend, I will clean up my repo so that it's ready for others to contribute to.

## Feb 11-17
As per the project requirements, I did not commit to or work on my project and only worked on a classmate's project.

## Feb 19
I realized that I was duplicating code in multiple functions, so I factored out code and wrote two general functions called `general_event_breakdown` and `process_data`. 

## Feb 20
I researched ways I can create a strike zone plotter that takes in data and plots the location of balls in the strike zone along with their outcome.

## Feb 22
I started creating the function and called it `batter_strike_zone_plotting`. I hope to also create an equivalent function for pitchers.

## Feb 23
The strike zone that my code was generating was behaving weirdly last night, as it was giving me a strike zone that looked like a wide rectangle instead of a tall rectangle. I'm still debugging the issue today.

## Feb 26
I finished creating the `batter_strike_zone_plotting` last night and today I'm making an equivalent function for pitchers.

## Feb 27 & 28
I added many more functions and put them into separate files. I also rearranged the code so all the necessary imports are in one file, updated the documentation in the repo, renamed functions because their names were too long, and added REFERENCES.md to cite my sources.


pyvis/data.py
# All imports needed for PyVis
from pybaseball import playerid_lookup
from pybaseball import statcast_batter
from pybaseball import statcast_pitcher
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import matplotlib.lines as lines

# Returns correct type of data (dependent on player and his type)
# Time can be a day or a range of dates (ex: process_data('2016-05-01', 'dustin', 'pedroia', 'batter') 
# or process_data('2016-05-01 to 2017-05-01', 'dustin', 'pedroia', 'batter'))
def process_data(time, firstname, lastname, pos):
    playerid = playerid_lookup(lastname, firstname)
    if "to" in time:
        if pos == 'batter': 
            return statcast_batter(time[0:10], time[14:], player_id=int(playerid['key_mlbam'][0]))
        elif pos == 'pitcher': 
            return statcast_pitcher(time[0:10], time[14:], player_id=int(playerid['key_mlbam'][0]))
        else: 
            return None
    else: 
        if pos == 'batter': 
            return statcast_batter(time, player_id=int(playerid['key_mlbam'][0]))
        elif pos == 'pitcher': 
            return statcast_pitcher(time, player_id=int(playerid['key_mlbam'][0]))
        else: 
            return None


pyvis/breakdown.py
import data # data.py

# BR in function names = breakdown (ex: batterBR = batter breakdown)

# Batter event breakdown in pie chart
# Time can be a day or a range of dates (ex: batterBR('2016-05-01', 'dustin', 'pedroia') 
# or batterBR('2016-05-01 to 2017-05-01', 'dustin', 'pedroia'))
def batterBR(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'batter')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    generalBR(data, 'hit event')
    
# Pitcher event breakdown in pie chart
# Time can be a day or a range of dates (ex: pitcherBR('2016-05-01', 'chris', 'sale') 
# or pitcherBR('2016-01-01 to 2017-01-01', 'chris', 'sale'))
def pitcherBR(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'pitcher')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    generalBR(data, 'pitch event')
    
# Pitch selection breakdown in pie chart
# Time can be a day or a range of dates (ex: pitchSelectionBR('2016-05-01', 'chris', 'sale') 
# or pitchSelectionBR('2016-01-01 to 2017-01-01', 'chris', 'sale'))
def pitchSelectionBR(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'pitcher')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    freq = data['pitch_type'].value_counts()
    print("Frequency of pitch breakdown:")
    print(freq)
    
    # Add to dict below as needed
    names = {'FF': '4 Seam Fastball', 'FA': '4 Seam Fastball', 'FT': '2 Seam Fastball', 'FC': 'Cutter', 
             'FS': 'Splitter', 'FO': 'Forkball', 'SI': 'Sinker', 'SL': 'Slider', 'CU': 'Curveball', 
             'KC': 'Knuckle Curve', 'EP': 'Eephus', 'CH': 'Changeup', 'SC': 'Screwball', 'KN': 'Knuckleball', 
             'IN': 'Unidentified', 'UN': 'Unidentified'}
    freq_edited = []
    for elm1, elm2 in zip(freq.keys().tolist(), freq.tolist()):
        if names[elm1] not in [r[0] for r in freq_edited]:
            freq_edited.append([names[elm1], elm2])
        else:
            for i, r in enumerate(freq_edited):
                if r[0] == names[elm1]:
                    freq_edited[i][1] += elm2
                    
    print("Pie chart of pitch selection:")
    plt.pie([row[1] for row in freq_edited], labels=[row[0] for row in freq_edited], autopct='%1.1f%%')
    plt.axis('equal')
    plt.tight_layout()
    plt.show()
    
# Takes in data from above functions and displays pie chart of data
def generalBR(data, name):
    data = data[np.isfinite(data['babip_value'])]
    freq = data['events'].value_counts()
    print("Frequency of " + name + " breakdown:")
    print(freq)
    
    print("Dropping infrequent events for pie chart.")
    total = sum(freq.tolist())
    freq_edited = [[elm1, elm2] for elm1, elm2 in zip(freq.keys().tolist(), freq.tolist()) if elm2 >= 0.02*total]
    other_total = [elm for elm in freq.tolist() if elm < 0.02*total]
    freq_edited.append(['other', sum(other_total)])
    
    print("Pie chart of " + name + " breakdown:")
    plt.pie([row[1] for row in freq_edited], labels=[row[0] for row in freq_edited], autopct='%1.1f%%')
    plt.axis('equal')
    plt.tight_layout()
    plt.show()


pyvis/piecharts.py
from pybaseball import playerid_lookup
from pybaseball import statcast_batter
import numpy as np
import matplotlib.pyplot as plt

# Hit breakdown
# Time can be a day or a range of dates (ex: hit_breakdown('2016-05-01', 'dustin', 'pedroia) 
# or hit_breakdown('2016-05-01 to 2017-05-01', 'dustin', 'pedroia'))
def hit_breakdown(time, firstname, lastname):
    playerid = playerid_lookup(lastname, firstname)
    print('Calculating hit breakdown. One moment please.')
    if "to" in time: data = statcast_batter(time[0:10], time[14:], player_id=int(playerid['key_mlbam'][0]))
    else: data = statcast_batter(time, player_id=int(playerid['key_mlbam'][0]))
    data = data[np.isfinite(data['babip_value'])]
    freq = data['events'].value_counts()
    print("Frequency of hit breakdown:")
    print(freq)
    print("Dropping infrequent events for pie chart.")
    total = sum(freq.tolist())
    freq_edited = [[elm1, elm2] for elm1, elm2 in zip(freq.keys().tolist(), freq.tolist()) if elm2 >= 0.02*total]
    other_total = [elm for elm in freq.tolist() if elm < 0.02*total]
    freq_edited.append(['other', sum(other_total)])
    print("Pie chart of hit breakdown:")
    plt.pie([row[1] for row in freq_edited], labels=[row[0] for row in freq_edited], autopct='%1.1f%%')
    plt.axis('equal')
    plt.tight_layout()
    plt.show()


pyvis/scatterplot.py
import data # data.py

# A dynamic scatterplot for a batter or pitcher, depending on user input
# Time can be a day or a range of dates (ex: scatterPlot("2017-09-30", "mookie", "betts", 'batter')
# or scatterPlot('2016-04-01 to 2016-05-01', 'chris', 'sale', 'pitcher'))
# Pos is either 'pitcher' or 'batter'
def scatterPlot(time, firstname, lastname, pos):
    data = processData(time, firstname, lastname, pos)
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    generalPlot(data, firstname, lastname)

# Generates a scatterplot based on user's input
def generalPlot(data, firstname, lastname):
    options = ['spin rate', 'release speed', 'launch speed', 'launch angle', 'hit distance', 'release extension']
    user = askUser(options)
    x_data = getRightData(data, int(user[0]))
    y_data = getRightData(data, int(user[1]))
    
    drop = []
    for i, (x, y) in enumerate(zip(x_data, y_data)):
        if np.isnan([x]) or np.isnan([y]):
            drop.append(i)
    x_data.drop(drop, inplace=True)
    y_data.drop(drop, inplace=True)
    
    print("Matching data points to type of pitch.")
    # get summary of pitches
    pitch_types = data.iloc[x_data.keys().tolist()]['pitch_type'].value_counts().keys().tolist()
    cmap = plt.get_cmap('viridis')
    # map a pitch to a specific color
    colors = cmap(np.linspace(0, 1, len(pitch_types)))
    
    names = {'FF': '4 Seam Fastball', 'FA': '4 Seam Fastball', 'FT': '2 Seam Fastball', 'FC': 'Cutter', 
             'FS': 'Splitter', 'FO': 'Forkball', 'SI': 'Sinker', 'SL': 'Slider', 'CU': 'Curveball', 
             'KC': 'Knuckle Curve', 'EP': 'Eephus', 'CH': 'Changeup', 'SC': 'Screwball', 'KN': 'Knuckleball', 
             'IN': 'Unidentified', 'UN': 'Unidentified'}
    allPatches = []
    for i, color in enumerate(colors):
        allPatches.append(patches.Patch(color=color, label=names[pitch_types[i]]))
    plt.scatter(x_data, y_data, c=colors)
    plt.xlabel(options[int(user[0])].title())
    plt.ylabel(options[int(user[1])].title())
    plt.title(options[int(user[1])].title() + " vs. " + options[int(user[0])].title() + " for " + firstname.title() + " " + lastname.title())

    print("Would you like a trendline (a.k.a. line of best fit) for this scatterplot? (yes/no)")
    trendline_decision = input()
    if trendline_decision.lower() in ["y", "yes"]:
        z = np.polyfit(x_data, y_data, 1)
        p = np.poly1d(z)
        plt.plot(x_data, p(x_data))
        print("y=%.6fx+(%.6f)"%(z[0],z[1]))
    plt.legend(handles=allPatches)
    plt.show()
    
# Asks user what two topics to plot
def askUser(options):
    print("Choose two topics to compare. Type the number next to the topic below, along with the axis you want it on.")
    print("For example, if you want to visualize spin rate (y-axis) vs. launch speed (x-axis), type '20' starting with \nthe x axis number.")
    for i, option in enumerate(options):
        if option == 'spin rate':
            print(i, option, '(rpm)')
        elif option == 'release speed':
            print(i, option, '(mph)')
        elif option == 'launch speed':
            print(i, option, '(mph)')
        elif option == 'launch angle':
            print(i, option, '(degrees)')
        elif option == 'hit distance':
            print(i, option, '(feet)')
        elif option == 'release extension':
            print(i, option, '(feet)')
    user = input()
    if user[0] == user[1]:
        print("You put the same topic twice. Try again.")
        user = input()
    return user
    
# Get data based on user's input
def getRightData(data, choice):
    if choice == 0:
        return data['release_spin_rate']
    elif choice == 1:
        return data['release_speed']
    elif choice == 2:
        return data['launch_speed']
    elif choice == 3:
        return data['launch_angle']
    elif choice == 4:
        return data['hit_distance_sc']
    elif choice == 5:
        return data['release_extension']
    else:
        return None


pyvis/spray_chart.py
import data # data.py

# Spray chart of a batter's hits
# Time can be a day or a range of dates (ex: sprayChart('2017-04-30', 'mookie', 'betts') 
# or sprayChart('2017-04-01 to 2017-05-01', 'mookie', 'betts'))
def sprayChart(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'batter')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    data = data[np.isfinite(data['babip_value'])]
    drawChart(data, firstname, lastname)

# Draw baseball field and plot data points
def drawChart(data, firstname, lastname):
    fig = plt.figure()
    print("Creating the spray chart.")
    # foul lines & base paths
    leftFoulLine = lines.Line2D((0,1), (1.5,0), color='black', transform=fig.transFigure, figure=fig)
    rightFoulLine = lines.Line2D((1,2), (0,1.5), color='black', transform=fig.transFigure, figure=fig)
    leftBaseLine = lines.Line2D((0.85,1), (0.225,0.45), color='gray', transform=fig.transFigure, figure=fig)
    rightBaseLine = lines.Line2D((1,1.15), (0.45,0.225), color='gray', transform=fig.transFigure, figure=fig)
    fig.lines.extend([leftFoulLine, rightFoulLine, leftBaseLine, rightBaseLine])
    
    # infield curve
    infieldPts = [(0.8,0.3), (1,0.7), (1.2,0.3)]
    infieldDir = [Path.MOVETO, Path.CURVE3, Path.CURVE3]
    infieldPath = Path(infieldPts, infieldDir)
    infield = patches.PathPatch(infieldPath, edgecolor='gray', facecolor='none', transform=fig.transFigure, lw=2)
    
    # outfield curve
    outfieldPts = [(0.35,0.975), (1,1.75), (1.65,0.975)]
    outfieldDir = [Path.MOVETO, Path.CURVE3, Path.CURVE3]
    outfieldPath = Path(outfieldPts, outfieldDir)
    outfield = patches.PathPatch(outfieldPath, edgecolor='gray', facecolor='none', transform=fig.transFigure, lw=2)
    fig.patches.extend([infield, outfield])
    
    # convert points to my coordinate system based on above values for infield & outfield points: my_x ==> [0.4, 1.6] & my_y ==> [0, 1.75]
    x_arr = [((((x-125)/120)+1)/(2/1.2))+0.4 for x in data['hc_x'] if not np.isnan(x)]
    y_arr = [((y-200)*-1)/(200/1.75) for y in data['hc_y'] if not np.isnan(y)]
    for x, y in zip(x_arr, y_arr):
        cir = patches.Circle((x,y), radius=.003, transform=fig.transFigure, color='blue')
        fig.patches.extend([cir])
    plt.figtext(0.8, 1.75, "Spray Chart of " + firstname.title() + " " + lastname.title(), size='large')
    plt.show()


pyvis/swing_plane.py
import data # data.py

# Swing plane of player's hits
# Time can be a day or a range of dates (ex: swingPlane('2017-04-30', 'mookie', 'betts') 
# or swingPlane('2017-04-01 to 2017-05-01', 'mookie', 'betts'))
def swingPlane(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'batter')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    indices = []
    for i, row in enumerate(data['description']):
        if row == 'hit_into_play_score' or row == 'hit_into_play_no_out': 
            indices.append(i)
    data = data.iloc[indices]
    data['launch_angle'] = data['launch_angle'].round()
    freq = data['launch_angle'].value_counts()
    drawPlane(freq, firstname, lastname)
        
# Draws swing plane based on given data from above
def drawPlane(data, firstname, lastname): 
    print("Generating swing plane for " + firstname.title() + " " + lastname.title() + ":")
    fig = plt.figure(figsize=(6,6))
    ax = fig.add_subplot(111, polar=True)
    ax.bar(data.keys().tolist(), data.tolist(), color='red', width=.075, linewidth=0.1, label='Hits')
    ax.set_thetamin(-80)
    ax.set_thetamax(80)
    ax.legend()
    plt.show()


pyvis/strike_zone.py
import data # data.py

# SZ in function names = strike zone (ex: batterSZ = batter strike zone)

# Plots strike zone of pitches for a batter
# Time can be a day or a range of dates (ex: batterSZ('2016-05-01', 'david', 'ortiz') 
# or batterSZ('2016-04-01 to 2016-05-01', 'david', 'ortiz'))
def batterSZ(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'batter')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    generalSZ(data, firstname, lastname, 'batter')

# Plots strike zone of pitches for a pitcher
# Time can be a day or a range of dates (ex: pitcherSZ('2017-05-30', 'chris', 'sale') 
# or pitcherSZ('2017-05-01 to 2017-06-01', 'chris', 'sale'))
def pitcherSZ(time, firstname, lastname):
    data = processData(time, firstname, lastname, 'pitcher')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    generalSZ(data, firstname, lastname, 'pitcher')

# Plots strike zone of pitches for a specific batter vs. pitcher matchup from the umpire's perspective
# Time can be a day or a range of dates (ex: matchupSZ("2016-08-01", "mookie", "betts", "james", "paxton")
# or matchupSZ("2016-08-01 to 2016-08-02", "mookie", "betts", "james", "paxton")
def matchupSZ(time, batter_firstname, batter_lastname, pitcher_firstname, pitcher_lastname):
    data = processData(time, batter_firstname, batter_lastname, 'batter')
    if data.size == 0: 
        return "Error: Did not find any data. Try again."
    pitcherid = playerid_lookup(pitcher_lastname, pitcher_firstname)
    data = data[data['pitcher'] == pitcherid['key_mlbam'][0]]
    generalSZ(data, batter_firstname, batter_lastname, 'batter')

# Takes in data from above functions and plots strike zone
def generalSZ(data, firstname, lastname, posid):
    print("Finding strike zone dimensions for player.")
    top = np.round(np.mean(data['sz_top']), 3) # height from ground to top of strike zone
    bot = np.round(np.mean(data['sz_bot']), 3) # height from ground to bottom of strike zone
    
    exclude = checkUser(data)
    data = fixData(data, exclude)
    labels = setLabels(exclude)
    
    print("Creating strike zone from player's perspective.")
    plt.figure(figsize=(6,6))
    ax = plt.subplot(111)
    if posid == 'batter': ax.scatter(data['plate_x'], data['plate_z'], c=setColors(data), marker='o')
    # default is x coords in batter's perspective, so multiply by -1 to reverse direction for pitcher
    elif posid == 'pitcher': ax.scatter([x_pos*-1 for x_pos in data['plate_x']], data['plate_z'], c=setColors(data), marker='o')
    # personalize strike zone per player
    ax.add_patch(patches.Rectangle((-20.14/24,bot), 20.14/12, top-bot, fill=False))
    ax.set_xlim(-2,2)
    ax.set_ylim(0,6)
    plt.legend(handles=labels)
    plt.xlabel("Horizontal Distance Away from Center of Plate (feet)")
    plt.ylabel("Height of Pitch (feet)")
    if posid == 'batter': plt.title("Strike Zone of " + firstname.title() + " " + lastname.title() + " from Umpire's Perspective")
    elif posid == 'pitcher': plt.title("Strike Zone of " + firstname.title() + " " + lastname.title() + " from Pitcher's Perspective")
    plt.show()

# Determine which options the user wants to see
def checkUser(data):
    options = ["hit into play", "called strikes", "swinging strikes & fouls", "balls"]
    exclude = []
    for option in options:
        print("Include pitches that are " + option + "? (yes/no)")
        ans = input()
        if ans.lower() in ["n", "no"]: 
            if option == "hit into play": 
                exclude.append("hit")
            elif option == "called strikes": 
                exclude.append("called_strike")
            elif option == "swinging strikes & fouls": 
                exclude.extend(["swinging_strike", "foul"])
            elif option == "balls": 
                exclude.append(option[0:len(option)-1])
    return exclude

# Remove unwanted data based on user's input
def fixData(data, exclude):
    options = data['description'].value_counts().keys().tolist()
    indices = []
    for i, r in enumerate(data['description']):
        for elm in exclude:
            if elm in r:
                indices.append(i)
    data.drop(data.index[indices], inplace=True)
    return data

# Set labels for graph
def setLabels(exclude):
    if "foul" in exclude: 
        exclude.pop(exclude.index("foul"))
        
    blue = patches.Patch(color='blue', label='Hit Into Play')
    red = patches.Patch(color='red', label='Called Strike')
    brown = patches.Patch(color='brown', label='Foul/Swinging Strike')
    green = patches.Patch(color='green', label='Ball')
    colors = [blue, red, brown, green]
    labels = [patch.get_label() for patch in colors]
    
    for elm in exclude:
        if elm == "hit": 
            colors.pop(labels.index("Hit Into Play"))
        elif elm == "called_strike": 
            colors.pop(labels.index("Called Strike"))
        elif elm == "swinging_strike": 
            colors.pop(labels.index("Foul/Swinging Strike"))
        elif elm == "ball": 
            colors.pop(labels.index("Ball"))
    return colors

# Set color depending on type of outcome on pitch
def setColors(data):
    colors = []
    for row in data['description']:
        if "hit" in row: 
            colors.append("blue")
        elif "called_strike" 
        in row: colors.append("red")
        elif "foul" in row or "swinging_strike" in row: 
            colors.append("brown")
        else: 
            colors.append("green")
    return colors
